File: LeetCode/Study Plan/Practice Problems/Medium/Maximum Length Of Concatenated String With Unique Characters/Go/maximum_length_of_concatenated_string_with_unique_characters.go
package main

import (
	"fmt"
	"strings"
)

func check_dup(str string) bool {
	n := len(str)
	mp := make(map[byte]int)
	for i := 0; i < n; i++ {
		mp[str[i]]++
		if mp[str[i]] > 1 {
			return true
		}
	}
	return false
}

func maximum_length_of_concatenated_string_with_unique_characters(arr []string) int {
	res := 0
	backtrack(arr, &res, 0, []string{})
	return res
}

func backtrack(arr []string, res *int, ind int, str []string) {
	n := len(arr)
	cnt := 0
	for i := ind; i < n; i++ {
		str = append(str, arr[i])
		val := strings.Join(str, "")
		if !check_dup(val) {
			cnt = len(val)
			backtrack(arr, res, i+1, str)
		}
		*res = max(*res, cnt)
		str = str[:len(str)-1]
	}
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	fmt.Println(maximum_length_of_concatenated_string_with_unique_characters([]string{"un", "iq", "ue"}))
}

File: LeetCode/Study Plan/Practice Problems/Medium/Maximum Length Of Concatenated String With Unique Characters/Cpp/maximum_length_of_concatenated_string_with_unique_characters.cc
#include <iostream>
#include <unordered_map>
#include <vector>

typedef std::vector<std::string> vs;
typedef std::unordered_map<char, int> uci;

bool check_dup(vs v);
void backtrack(vs v, size_t *res, size_t ind, vs temp);
std::string join_string(vs v);

int maximum_length_of_concatenated_string_with_unique_characters(vs &v) {
    size_t res =0;
    backtrack(v, &res, 0, vs{});
    return res;
}

bool check_dup(vs v) {
    uci um;
    for (auto itr = v.begin(); itr!=v.end(); itr++) {
        std::string word = *itr;
        for (int i=0; i<word.size(); i++) {
            um[word[i]]++;
            if (um[word[i]] >1 ) return true;
        } 
    }
    return false;
}

void backtrack(vs v, size_t *res, size_t ind, vs temp) {
    size_t cnt = 0;
    for (int i=ind; i<v.size(); i++){
        temp.push_back(v[i]);
        std::string word = join_string(temp);
        if (!check_dup(temp)) {
            cnt = word.size();
            backtrack(v, res, i+1, temp);
        }
        *res = std::max(*res, cnt);
        temp.pop_back();
    }
}

std::string join_string(vs v) {
    std::string s = "";
    for (auto itr = v.begin(); itr!=v.end(); itr++) s += *itr;
    return s;
}

int main() {
    vs v{"un","iq","ue"};
    std::cout<<maximum_length_of_concatenated_string_with_unique_characters(v)<<"\n";
}
